src/library/scala/Array.scala

```diff
@@ -238,10 +238,13 @@ object Array extends FallbackArrayBuilding {
    *  @return   the array created from concatenating `xss`
    */
   def concat[T: ClassTag](xss: Array[T]*): Array[T] = {
-    val b = newBuilder[T]
-    b.sizeHint(xss.map(_.length).sum)
-    for (xs <- xss) b ++= xs
-    b.result()
+    if (xss.foldLeft(0)(_ + _.length) == 0) empty[T]
+    else {
+      val b = newBuilder[T]
+      b.sizeHint(xss.map(_.length).sum)
+      for (xs <- xss) b ++= xs
+      b.result()
+    }
```

Comment:
What about

```scala
val totalLength = xss.map(_.length).sum
if (totalLength == 0) empty[T]
else {
  val b = newBuilder[T]
  b.sizeHint(totalLength)
  for (xs <- xss) b ++= xs
  b.result()
}
```

to save on the `foldLeft`?

Reply:
I think I will rewrite it to simple while loop to avoid allocation altogether.

Reply:
Out of curiosity I compared efficiency of different summing methods:

```
[info] ListBenchmark.profile_sum_fold                    0  avgt   40   165,869 ±  10,580  ns/op
[info] ListBenchmark.profile_sum_fold                    1  avgt   40   163,242 ±  10,485  ns/op
[info] ListBenchmark.profile_sum_fold                   10  avgt   40   209,199 ±  12,308  ns/op
[info] ListBenchmark.profile_sum_fold                  100  avgt   40   806,816 ±  36,106  ns/op
[info] ListBenchmark.profile_sum_fold                 1000  avgt   40  6074,634 ±  92,882  ns/op
[info] ListBenchmark.profile_sum_foreach                 0  avgt   40   161,354 ±  13,025  ns/op
[info] ListBenchmark.profile_sum_foreach                 1  avgt   40   160,287 ±  10,815  ns/op
[info] ListBenchmark.profile_sum_foreach                10  avgt   40   199,170 ±  15,542  ns/op
[info] ListBenchmark.profile_sum_foreach               100  avgt   40   748,299 ±  33,686  ns/op
[info] ListBenchmark.profile_sum_foreach              1000  avgt   40  6355,685 ±  81,311  ns/op
[info] ListBenchmark.profile_sum_map                     0  avgt   40   213,734 ±  12,799  ns/op
[info] ListBenchmark.profile_sum_map                     1  avgt   40   222,675 ±  14,131  ns/op
[info] ListBenchmark.profile_sum_map                    10  avgt   40   253,827 ±  17,178  ns/op
[info] ListBenchmark.profile_sum_map                   100  avgt   40   846,869 ±  30,010  ns/op
[info] ListBenchmark.profile_sum_map                  1000  avgt   40  6487,538 ± 102,901  ns/op
```